src/main/scala/workflow/Optimizer.scala

```diff
@@ -0,0 +1,37 @@
+package workflow
+
+/**
+ * Optimizes a Pipeline DAG
+ */
+object Optimizer extends RuleExecutor {
+  protected val batches: Seq[Batch] = Batch("DAG Optimization", FixedPoint(100), EquivalentNodeMerger) :: Nil
```

Comment:
@tomerk can correct me - but this is basically moving pipelines from a tree structure with a lot of redundant branches into a more compact DAG that eliminates much of this redundancy. Execution of the pipeline is basically memo-ized depth-first search from the result in this DAG, so consolidating the representation helps a bunch!